runtime/src/threads/pthreads/threads-pthreads.c

```diff
@@ -297,14 +475,35 @@ void chpl_thread_exit(void) {
 
   pthread_mutex_unlock(&thread_info_lock);
 
+  pthread_mutex_lock(&thread_stack_list_lock);
+
   while (thread_list_head != NULL) {
-    if (pthread_join(thread_list_head->thread, NULL) != 0)
-      chpl_internal_error("thread join failed");
     tlp = thread_list_head;
+    if (pthread_join(tlp->thread, NULL) != 0)
+      chpl_internal_error("thread join failed");
+
+    /* We have joined with tlp->thread, so now we search for its stack
+     * in the thread_stack_list_head list and free it. This is for be
+     * sure that we free stacks used only by teminated threads.
+     */
+    for (tslp = thread_stack_list_head; tslp != NULL; tslp = tslp->next) {
+      if(pthread_equal(tlp->thread, tslp->owner_pthread)){
+          chpl_free_pthread_stack(tslp->stack);
+          if(tslp->prec == NULL)
+            thread_stack_list_head = tslp->next;
+          else
+            tslp->prec->next = tslp->next;
+          chpl_mem_free(tslp, 0, 0);
```

Comment:
tslp->next->prec needs updating here also, or the list will be broken and a future operation on tslp->next will be working with a prec pointer which points to the (freed) tslp we are operating on now.